File: cpp_src/DivideTwoIntegers.cpp
/**
 * -2147483648, -1
 * overflow
 */
class Solution {
public:
    int divide(int dividend, int divisor) {
        if(divisor == 0) return INT_MAX;
        bool negative = false;
        long dividendL = (long)dividend;
        long divisorL = (long)divisor;
        if((dividendL > 0 && divisorL < 0) || (dividendL < 0 && divisorL > 0)) negative = true;
        if(dividendL < 0) dividendL = -dividendL;
        if(divisorL < 0) divisorL = -divisorL;
        long result = 0;
        long temp = 0;
        do{
            result += expand(dividendL,divisorL,temp);
            dividendL = temp;
        }while(dividendL >= divisorL);
        if(negative) return (int)-result;
        else{
            if(result > INT_MAX) return INT_MAX;
            else return (int)result;
        }
    }
    
    //each time, we minus divisor^n, where n is the max integer that satisfies divisor^n < dividend,
    //the factor should be 2^n.
    int expand(long dividend, long divisor,long& mod){
        if(dividend < divisor){
            return 0;
        }
        int n = 0;
        while(dividend >= divisor){
            n++;
            divisor = divisor << 2;
        }
        mod = dividend - (divisor >> 2);
        n--;
        int result = 1;
        for(int i = 0; i < n; i++){
            result = result << 2;
        }
        return result;
    }
};

File: python_src/DivideTwoIntegers.py
class Solution(object):
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        is_positive = True
        if dividend < 0 and divisor < 0: 
            dividend = -dividend #may overflow, but in python, it's ok
            divisor = -divisor
        elif dividend > 0 and divisor < 0:
            divisor = -divisor
            is_positive = False
        elif dividend < 0 and divisor > 0:
            dividend = -dividend
            is_positive = False
        result = 0
        n = 0
        while dividend >= divisor:
            divisor <<= 1
            n += 1
        if n <= 0: return 0
        factor = 1
        for i in range(n):
            factor <<= 1
        
        while n >= 0:
            if dividend >= divisor:
                result += factor
                dividend -= divisor
            divisor >>= 1
            factor >>= 1
            n -= 1
        if not is_positive: result = -result
        if result > 0x7fffffff: return 0x7fffffff
        return result
            
            
            
        